### ISSUE TITLE
Option to not flush cookies between tests

### ISSUE THREAD

**Issue by igorpavlov** [cypress-io/cypress#959]
- Operating System: OSX 10.12.6 (16G1036)
- Cypress Version: 1.1.1
- Browser Version:  62.0.3202.94

### Is this a Feature or Bug?

Unexpected bahaviour.

### Current behavior:

Cookies are automatically flushed between tests (as per the documentation).

![image](https://user-images.githubusercontent.com/3973243/33242300-77d3e2aa-d2ca-11e7-87d1-eb9e81240439.png)

### Desired behavior:

Have an option to not flush cookies between tests for more control.

### How to reproduce:

1. Write a first test that sets a cookie to the domain;
2. Write a second test that should utilise this cookie;
3. The cookie from the first test is flushed and is not available in the second test.

#### Test code:

```js
describe('Set cookie', function() {
  it('Sets cookie', function() {
    cy.setCookie('test', '1');
  });
});

describe('Get cookie', function() {
  it('Gets cookie', function() {
    cy.getCookie('test').should('have.property', 'value', '1')
  });
});
```

### Additional Info (images, stack traces, etc)

![image](https://user-images.githubusercontent.com/3973243/33242272-d92e9a1e-d2c9-11e7-9d06-46eeba935432.png)

Use case: test user to sign up and set a password.

![444](https://user-images.githubusercontent.com/3973243/33242202-f2a23e98-d2c8-11e7-8fcd-b6527db3ad2f.gif)

--- 
Temporary solution:

1) Once logged in, save session cookie in a variable
2) Wrap `it` into another function that will at first set that cookie back and then run events and assertions
```js
var sessionId = '';

it('signs up', function() {
  cy.get('.sign-up')
    .click();
  cy.getCookie('connect.sid').then(function (session) {
    sessionId = session.value;
  });
});

IT('does something after sign up', function() {
  // Session is now initialised for this test
});

/* Helpers */

function initSession () {
  cy.setCookie('connect.sid', sessionId);
}

function IT (does, assert) {
  it(does, function() {
    initSession();
    assert();
  });
}
```

**Comment by brian-mann** [comment]
There is an API to preserve cookies in between tests: https://docs.cypress.io/api/cypress-api/cookies.html

I'm closing this issue in favor of this one (below) because it goes in length to explain why what we do by default is a bad idea and what needs to change. It's a priority but its underneath other things unfortunately.

https://github.com/cypress-io/cypress/issues/686

**Comment by igorpavlov** [comment]
Thanks @brian-mann . Hopefully my "hack" would help some people as it actually works for me and my usecase.

**Comment by xiangweilee** [comment]
This will preserve cookies in every test
```js
beforeEach(() => {
  // Preserve cookie in every test
  Cypress.Cookies.defaults({
    whitelist: (cookie) => {
      return true;
    }
  })
});

```

**Comment by trentrand** [comment]
For those arriving from the future, starting in version 5.0.0 the `Cypress.Cookies.defaults()` `whitelist` option has been renamed to `preserve` to more closely reflect its behavior. Addressed in #7782.

https://docs.cypress.io/guides/references/changelog.html#5-0-0

**Comment by pgarciacamou** [comment]
I ended up using `Cypress.Cookies.preserveOnce` instead. Unfortunately, it doesn't take in a function or a RegExp like `Cypress.Cookies.defaults` does, but I managed to work around it:

```js
  cy.getCookies().then(cookies => {
    const cookiesNames = cookies.map(c => c.name)
    cy.log(JSON.stringify({cookies: cookiesNames}))
    Cypress.Cookies.preserveOnce(...cookiesNames)
  })
```

Hopefully, this will help someone else.

---------

### Explanation

The problem with `Cypress.Cookies.defaults`, like in 

> This will preserve cookies in every test
> 
> ```js
> beforeEach(() => {
>   // Preserve cookie in every test
>   Cypress.Cookies.defaults({
>     whitelist: (cookie) => {
>       return true;
>     }
>   })
> });
> ```
> see [this comment](https://github.com/cypress-io/cypress/issues/959#issuecomment-659407567)



Is that according to the docs (https://docs.cypress.io/api/cypress-api/cookies#Defaults), `Cypress.Cookies.defaults` will preserve all cookies for the rest of the execution of all tests.

> You can modify the global defaults and preserve a set of Cookies which will **always** be preserved across tests.
> Any change you make here will take effect immediately for the remainder of **every single** test.
> A great place to put this configuration is in your cypress/support/index.js file, since it is loaded **before any test files** are evaluated.


In other words, according to the docs, using `Cypress.Cookies.defaults` at the `beforeEach` level is wrong. And what is worse, it will affect the rest of the tests (all of them).

**Comment by ganeshsirsi** [comment]
You can also look into this code, This reads and preserves all cookies, If you are not particular about something you can use this
[](https://ganeshsirsi.medium.com/cypress-how-to-preserve-cookies-and-keep-login-session-active-in-each-test-simplest-way-717bbc11adf7)

**Comment by mekalag** [comment]
Guys, we just started writing our automation in cypress, we will most require to store cookies to be used in all the spec files except for one or two specs say for ex: when testing authentication section.

So added : 
```
Cypress.Cookies.defaults({
  preserve: ['auth', 'TOKEN'],
});
```
in cypress/support/index.js file.

How can I exclude few files to not use cookies? Please suggest

**Comment by pgarciacamou** [comment]
@mekalag that's separate to this issue but I wanted to add a suggestion in case it helps other people end up in this thread looking for an answer to your question. 

One _userland_ idea I had is to clear the cookies for the command you need to run and then restoring them:

```js
/**
 * Clears cookies before executing a callback and then restores the cookies.
 * @param {Function} callback
 * @param {{ domain, log, timeout }} options https://docs.cypress.io/api/commands/getcookies#Arguments
 */
Cypress.Commands.add('ignoreCookiesOnce', (callback, options) => {
  return cy.getCookies(options).then(cookies => {
    // Clear cookies
    cy.clearCookies(options)

    // Execute callback
    callback()

    // Restore cookies
    cookies.forEach(({name, value, ...rest}) => {
      cy.setCookie(name, value, rest)
    })
  })
})
```

Then you could do:

```js
cy.ignoreCookiesOnce(
  () => {
    cy.request({
      url: 'https://...',
      method: 'POST',
      headers: {
        Cookie: 'whateveryouwant'
      },
      body: `...`
    }).then(({body}) => body)
  },
  {domain: 'yourdomain.com'} // optional
)
```

Hope this helps 👍 

-------------

You can also be much more flexible if you need to pass specific cookies to clear:

```js
/**
 * Clears cookies before executing a callback and then restores the cookies.
 * Usage:
 *  1. cy.ignoreCookiesOnce(() => cy.visit(...))
 *  2. cy.ignoreCookiesOnce(() => cy.request(...), { domain: 'yourdomain.com'})
 *  3. cy.ignoreCookiesOnce(() => cy.visit(...), { cookies: ['foo', 'bar']})
 *  4. cy.ignoreCookiesOnce(() => cy.request(...), { except: ['foo', 'bar']})
 * @param {Function} callback
 * @param {{ cookies: Array, except: Array, domain: string, log, timeout }} options See: https://docs.cypress.io/api/commands/getcookies#Arguments
 */
Cypress.Commands.add('ignoreCookiesOnce', (callback, options = {}) => {
  Cypress.Cookies.debug(true, {verbose: true})
  let {cookies, except, ...cyOptions} = options

  return cy.getCookies(cyOptions).then((all = []) => {
    // Obtain cookies from options
    if (Array.isArray(cookies) && cookies.length > 0) {
      cookies = cookies.map(n => all.find(c => c.name === n))
    } else if (Array.isArray(except) && except.length > 0) {
      cookies = all.filter(c => !except.includes(c.name))
    } else {
      cookies = all
    }

    // Clear cookies
    if (cookies === all) {
      cy.clearCookies(cyOptions) // clear all
    } else {
      cookies.forEach(c => cy.clearCookie(c.name, cyOptions))
    }

    // Execute callback
    callback()

    // Restore cookies
    cookies.forEach(({name, value, ...rest}) => {
      cy.setCookie(name, value, rest)
    })
  })
})
```

I also added this answer to stackoverflow: https://stackoverflow.com/a/75035723/4984618.
